### PR TITLE
Assignment4

### PR DESCRIPTION
**Pull request by jstanley6**

### CODE REVIEW

File: Assignment4/Logger.Tests/StorageTests.cs [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        Student studentEntity = new Student()
+        {
+            Id = Guid.NewGuid(),
+            FullName = new FullName("Jason", "Stanley"),
+            StudentYear = StudentClassification.Senior
+
+        };
+        storage.Add(studentEntity);
+        storage.Remove(studentEntity);
+        Assert.IsFalse(storage.Contains(studentEntity));
```
Comment (j0shh3ss) [inline Assignment4/Logger.Tests/StorageTests.cs:101]: Add tests for if inputs are null

File: Assignment4/Logger/bin/Debug/net7.0/Logger.deps.json [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    }
+  },
+  "libraries": {
+    "Logger/1.0.0": {
+      "type": "project",
+      "serviceable": false,
+      "sha512": ""
+    }
+  }
+}
```
Comment (j0shh3ss) [inline Assignment4/Logger/bin/Debug/net7.0/Logger.deps.json:23]: Comments explaining why you added this would be helpful.

**Review by j0shh3ss — COMMENTED** [review]
Define a full name record (first, last, middle) handling optional and null appropriately. ✔
Provide a comment on the full name record on why you selected to define a value or a reference type and ✔
Provide a comment on the full name record on why or why not the type is immutable. ✔
Define an IEntity interface: ✔
Add an Id property of type Guid that is an init-only setter. ✔
Add a Name property that is string. ✔
Define an abstract base class that implements IEntity - appropriately choosing to implement the interface explicitly or implicitly. ✔
Do not implement the Name property in this abstract class. ❌
Do force any derived classes to provide an implementation for Name. ✔
Define book, student, and employee records - all with entity capabilities. ✔
Choose a thoughtful approach for the value of Name for each type of entity. In all cases, there should be no backing field for Name (not even one generated by an automatically implemented property) - consider a calculated property. ✔
Provide a comment on each interface member in each entity explaining why you implemented it implicitly or explicitly. ✔
You should consider the relationship between Student and Employee and refactor the common code shared between them. ✔
Given the properties on the full name record, you should consider which entities it makes sense to use it. ✔
Test that the equality behavior on these entities behaves as expected. ✔
Provide a comment on each interface method explaining why you implemented it implicitly or explicitly. ❌
Unit test the Storage class using various entities to ensure that there are no bugs.
Code review (using TODO: comments if there are any issues - which there may not be) all logger-related classes.) ❌
Be sure you enable:
Enable Nullability for all projects ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Ensure there are no errors or warnings (including code analysis warnings) ✔
All of the above should be unit tested ❌
Choose simplicity over complexity ❌
^^^ This last question you need to focus on. I found it extremely hard to grade due to how many files you edited, and it made it way too complex. The main portion of the lab looks pretty good, but I see a lot of unnecessary changes made in documents without comments explaining why.

File: Assignment4/Logger/IEntity.cs [hunk 3]
```diff
@@ -1,7 +1,5 @@
 ﻿namespace Logger;
 public interface IEntity
 {
-
```
Comment (JessicaH46) [inline Assignment4/Logger/IEntity.cs:4]: Still missing the guid id and string name members here, I'm not sure if you wanted to leave it empty as you edited it here

File: Assignment4/Logger.Tests/bin/Debug/net7.0/Logger.Tests.runtimeconfig.json [hunk 4]
```diff
@@ -0,0 +1,9 @@
+{
```
Comment (JessicaH46) [inline Assignment4/Logger.Tests/bin/Debug/net7.0/Logger.Tests.runtimeconfig.json:1]: I am a little confused as to what this part of your code is here, is this supposed to be for the fundamentals of the assignment? I haven't seen anyone do it like this, usually its in this format if you were doing it for the fundamentals.

`<EnableNETAnalyzers>true</EnableNETAnalyzers>`

File: Assignment4/Logger/BookStudentEmployeeRecords.cs [hunk 5]
```diff
@@ -0,0 +1,27 @@
+﻿using System;
```
Comment (JessicaH46) [inline Assignment4/Logger/BookStudentEmployeeRecords.cs:1]: I think this file looks good here, we also did records for ours, good job!

**Review by JessicaH46 — COMMENTED** [review]
I think overall from what I could see you have most of the key points done for the assignment. Good job! However, there were a lot of edited files that I don't think were needed and they just became cluster for your code and made it difficult to navigate. I think you could probably remove a lot of the unnecessary files to simplify your code for this assignment to meet the following requirement.
-Choose simplicity over complexity 
I think you also need to ass some NullArgumentException in your tests to check for null. Good work on the code! Just might be overcomplicating it for some parts.

File: Assignment4/Logger.Tests/bin/Debug/net7.0/Logger.Tests.runtimeconfig.json (hunk 4 above)
Comment (jstanley6) [reply Assignment4/Logger.Tests/bin/Debug/net7.0/Logger.Tests.runtimeconfig.json:1]: My GitHub got messed up and the .gitIgnore got deleted and I almost lost the whole project and had to redo configurations to get it to build again and since the .gitIgnore was gone it pushed all of it up. I think I messed up when I created the branch, I have to use TortoiseGit to push any changes I make up to Github. I will work on fixing it all after this assignment is reviewed incase I have to delete all projects and just redo the whole repo

File: Assignment4/Logger/bin/Debug/net7.0/Logger.deps.json (hunk 2 above)
Comment (jstanley6) [reply Assignment4/Logger/bin/Debug/net7.0/Logger.deps.json:23]: My GitHub got messed up and the .gitIgnore got deleted and I almost lost the whole project and had to redo configurations to get it to build again and since the .gitIgnore was gone it pushed all of it up. I think I messed up when I created the branch, I have to use TortoiseGit to push any changes I make up to Github. I will work on fixing it all after this assignment is reviewed incase I have to delete all projects and just redo the whole repo

File: Assignment4/Logger.Tests/StorageTests.cs (hunk 1 above)
Comment (jstanley6) [reply Assignment4/Logger.Tests/StorageTests.cs:101]: Why does this have a red mark?
Do not implement the Name property in this abstract class. ❌
Did I miss something here?

**Comment by jstanley6** [comment]
Updated assignment per specifications. Although there is still files not needed that were pushed up, I almost deleted the whole assignment and since it is already due and I might need to redo it it would show up as late, but the other problems/suggestions have been fixed

File: Assignment4/Logger/FullName.cs [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+//immutable objects never change so therefore can't be altered.
+
+public record struct FullName(string FirstName, string LastName, string? MiddleName = null)
+{
+
+    public string FirstName { get; } = FirstName ?? throw new ArgumentNullException(nameof(FirstName));
+    public string LastName { get; } = LastName ?? throw new ArgumentNullException(nameof(LastName));
+    public string? MiddleName { get; } = MiddleName;
+
+    public string Name => $"{FirstName} {MiddleName} {LastName}";
```
Comment (mmwoodfo) [inline Assignment4/Logger/FullName.cs:20]: Looks good, you could end up with 2 spaces here if middle name is null!

File: Assignment4/Logger/BookStudentEmployeeRecords.cs [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+namespace Logger;
+
+public record Book : IEntity
```
Comment (mmwoodfo) [inline Assignment4/Logger/BookStudentEmployeeRecords.cs:5]: These records should be split into their own files

Comment (mmwoodfo) [inline Assignment4/Logger/BookStudentEmployeeRecords.cs:5]:
```suggestion
public record Book : BaseEntity
```

File: Assignment4/Logger.Tests/TestEntities.cs [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+        Student student2 = new Student()
+        {
+            Id = student1.Id,
+            FullName = student1.FullName,
+            StudentYear = student1.StudentYear
+        };
+        Assert.AreEqual(student1, student2);
+    }
+    
+    [TestMethod]
```
Comment (mmwoodfo) [inline Assignment4/Logger.Tests/TestEntities.cs:26]: Also check for inequality!

**Review by mmwoodfo — COMMENTED** [review]
I think you have so many changes because you update all the file locations.
